File: tools/os-test/run_verify_os_server_full_wrapper.py
#!/usr/bin/env python3
"""
Start Floorp Debug (`deno task feles-build dev`), wait for the OS server health
endpoint to come up, run `verify_os_server_full.py`, and tear everything down.
"""

from __future__ import annotations

import argparse
import os
import signal
import subprocess
import sys
import time
from pathlib import Path

ROOT_DIR = Path(__file__).resolve().parents[2]
DEFAULT_CMD = ["deno", "task", "feles-build", "dev"]


def wait_for_health(timeout: float, interval: float) -> bool:
    from verify_os_server_full import make_request  # type: ignore

    deadline = time.time() + timeout
    while time.time() < deadline:
        status, _ = make_request("/health")
        if status == 200:
            return True
        time.sleep(interval)
    return False


def stop_process(proc: subprocess.Popen[str], grace: float = 10.0) -> None:
    if proc.poll() is not None:
        return
    try:
        os.killpg(proc.pid, signal.SIGTERM)
    except Exception:
        try:
            proc.terminate()
        except Exception:
            pass
    try:
        proc.wait(timeout=grace)
        return
    except subprocess.TimeoutExpired:
        pass
    try:
        os.killpg(proc.pid, signal.SIGKILL)
    except Exception:
        try:
            proc.kill()
        except Exception:
            pass
    try:
        proc.wait(timeout=3)
    except Exception:
        pass


def run_verifier() -> int:
    import verify_os_server_full as verifier

    try:
        verifier.main()
        return 0
    except SystemExit as exc:  # noqa: BLE001
        code = exc.code
        if code is None:
            return 0
        return int(code)


def main(argv: list[str] | None = None) -> int:
    parser = argparse.ArgumentParser(description="Run Floorp OS server verification end-to-end")
    parser.add_argument(
        "--cmd",
        nargs=argparse.REMAINDER,
        help="Custom command to start Floorp Debug (default: deno task feles-build dev)",
    )
    parser.add_argument(
        "--timeout",
        type=float,
        default=120.0,
        help="Seconds to wait for /health to respond",
    )
    parser.add_argument(
        "--interval",
        type=float,
        default=2.0,
        help="Seconds between health checks",
    )
    parser.add_argument(
        "--skip-start",
        action="store_true",
        help="Do not start Floorp Debug; assume it is already running",
    )

    args = parser.parse_args(argv)

    cmd = args.cmd if args.cmd else DEFAULT_CMD
    proc: subprocess.Popen[str] | None = None

    if args.skip_start:
        print("Skipping Floorp Debug start; assuming server is already running.")
    else:
        print(f"Starting Floorp Debug: {' '.join(cmd)} (cwd={ROOT_DIR})")
        proc = subprocess.Popen(
            cmd,
            cwd=ROOT_DIR,
            stdin=subprocess.DEVNULL,
            stdout=None,
            stderr=None,
            text=True,
            start_new_session=True,
        )

        print("Waiting for OS server health endpoint...")
        if not wait_for_health(args.timeout, args.interval):
            print("OS server did not become healthy within timeout.", file=sys.stderr)
            stop_process(proc)
            return 1
        print("OS server is healthy; waiting 20s for full startup...")
        time.sleep(20)
        print("Proceeding with verification.")

    exit_code = run_verifier()

    if proc is not None:
        print("Stopping Floorp Debug...")
        stop_process(proc)

    return exit_code


if __name__ == "__main__":
    sys.exit(main())

File: tools/os-test/run_verify_os_server_full_wrapper.ts
#!/usr/bin/env -S deno run -A
// SPDX-License-Identifier: MPL-2.0

import { parseArgs } from "@std/cli/parse-args";

const DEFAULT_BASE_URL = Deno.env.get("FLOORP_OS_BASE_URL") ?? "http://127.0.0.1:58261";
const DEFAULT_START_CMD = ["deno", "task", "feles-build", "dev"];

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

async function waitForHealth(baseUrl: string, timeoutMs: number, intervalMs: number): Promise<boolean> {
  const deadline = Date.now() + timeoutMs;
  const url = `${baseUrl.replace(/\/$/, "")}/health`;

  while (Date.now() < deadline) {
    try {
      const response = await fetch(url, { signal: AbortSignal.timeout(Math.max(intervalMs, 2_000)) });
      if (response.status === 200) {
        return true;
      }
    } catch {
      // retry
    }
    await sleep(intervalMs);
  }

  return false;
}

async function stopProcess(process: Deno.ChildProcess): Promise<void> {
  if (process.pid === undefined) {
    return;
  }

  if (Deno.build.os === "windows") {
    await new Deno.Command("taskkill", {
      args: ["/PID", String(process.pid), "/T", "/F"],
      stdout: "null",
      stderr: "null",
    }).output().catch(() => undefined);
    return;
  }

  try {
    process.kill("SIGTERM");
  } catch {
    // ignore
  }

  await Promise.race([
    process.status,
    sleep(10_000),
  ]);

  try {
    process.kill("SIGKILL");
  } catch {
    // ignore
  }
}

function splitVerifierArgs(args: string[]): { wrapperArgs: string[]; verifierArgs: string[] } {
  const separatorIndex = args.indexOf("--");
  if (separatorIndex < 0) {
    return { wrapperArgs: args, verifierArgs: [] };
  }
  return {
    wrapperArgs: args.slice(0, separatorIndex),
    verifierArgs: args.slice(separatorIndex + 1),
  };
}

export async function main(argv: string[] = Deno.args): Promise<number> {
  const { wrapperArgs, verifierArgs } = splitVerifierArgs(argv);

  const parsed = parseArgs(wrapperArgs, {
    string: ["base-url", "timeout", "interval", "cmd"],
    boolean: ["skip-start"],
    default: {
      "base-url": DEFAULT_BASE_URL,
      timeout: "120",
      interval: "2",
      "skip-start": false,
    },
  });

  const baseUrl = String(parsed["base-url"]);
  const timeoutMs = Math.max(1, Number(parsed.timeout)) * 1000;
  const intervalMs = Math.max(1, Number(parsed.interval)) * 1000;
  const skipStart = parsed["skip-start"] === true;

  Deno.env.set("FLOORP_OS_BASE_URL", baseUrl);

  const startCommand = typeof parsed.cmd === "string" && parsed.cmd.trim().length > 0
    ? parsed.cmd.trim().split(/\s+/)
    : DEFAULT_START_CMD;

  let floorpProcess: Deno.ChildProcess | null = null;

  try {
    if (skipStart) {
      console.log("Skipping Floorp start; expecting server already running.");
    } else {
      console.log(`Starting Floorp: ${startCommand.join(" ")}`);
      floorpProcess = new Deno.Command(startCommand[0], {
        args: startCommand.slice(1),
        cwd: Deno.cwd(),
        stdin: "null",
        stdout: "inherit",
        stderr: "inherit",
      }).spawn();

      console.log("Waiting for OS API health endpoint...");
      if (!(await waitForHealth(baseUrl, timeoutMs, intervalMs))) {
        console.error("OS API server did not become healthy within timeout.");
        await stopProcess(floorpProcess);
        return 1;
      }

      console.log("OS API health is ready; waiting 20s for full startup...");
      await sleep(20_000);
    }

    const verifyArgs = [
      "run",
      "-A",
      "tools/os-test/verify_os_server_full.ts",
      "--base-url",
      baseUrl,
      ...verifierArgs,
    ];

    const verify = new Deno.Command(Deno.execPath(), {
      args: verifyArgs,
      cwd: Deno.cwd(),
      stdout: "inherit",
      stderr: "inherit",
    }).spawn();

    const status = await verify.status;
    return status.code;
  } finally {
    if (floorpProcess) {
      console.log("Stopping Floorp...");
      await stopProcess(floorpProcess);
    }
  }
}

if (import.meta.main) {
  const code = await main(Deno.args);
  Deno.exit(code);
}
